execute/dispatcher.go

```diff
@@ -107,14 +107,24 @@ func (d *poolDispatcher) setErr(err error) {
 
 //Stop the dispatcher.
 func (d *poolDispatcher) Stop() error {
+	// Check if we are already closed and return the error if so.
 	d.mu.Lock()
-	defer d.mu.Unlock()
 	if d.closed {
+		defer d.mu.Unlock()
 		return d.err
 	}
+	// If not, mark the dispatcher as closed and signal to the current
+	// workers that they should stop processing more work.
 	d.closed = true
 	close(d.closing)
+	d.mu.Unlock()
+
+	// Wait for the existing workers to finish.
 	d.wg.Wait()
```

Comment:
Does waiting have a chance to set the err? Does this mean that only the first caller to Stop is guaranteed to get the real error? Since multiple calls to Stop don't wait for the error to be set do we now risk loosing the error?